File: CC/tests/longest_substr_test.cc
#include "../src/longest_substr.hpp"
#include <gtest/gtest.h>

class LongestSubstrTest : public ::testing::Test {
protected:
  Solution solution;
};

TEST_F(LongestSubstrTest, basic_cases) {
  EXPECT_EQ(solution.lengthOfLongestSubstring("abcabcbb"), 3);
  EXPECT_EQ(solution.lengthOfLongestSubstring("bbbbb"), 1);
  EXPECT_EQ(solution.lengthOfLongestSubstring("pwwkew"), 3);
}

TEST_F(LongestSubstrTest, edge_cases) {
  EXPECT_EQ(solution.lengthOfLongestSubstring(""), 0);
  EXPECT_EQ(solution.lengthOfLongestSubstring(" "), 1);
  EXPECT_EQ(solution.lengthOfLongestSubstring("au"), 2);
  EXPECT_EQ(solution.lengthOfLongestSubstring("dvdf"), 3);
}

TEST_F(LongestSubstrTest, special_characters) {
  EXPECT_EQ(solution.lengthOfLongestSubstring("!@#!!@"), 3);
  EXPECT_EQ(solution.lengthOfLongestSubstring("123123123"), 3);
  EXPECT_EQ(solution.lengthOfLongestSubstring("a1b2c3"), 6);
}

TEST_F(LongestSubstrTest, longer_strings) {
  EXPECT_EQ(solution.lengthOfLongestSubstring("abcdefg"), 7);
  EXPECT_EQ(solution.lengthOfLongestSubstring("ababab"), 2);
  EXPECT_EQ(solution.lengthOfLongestSubstring("abba"), 2);
}

TEST_F(LongestSubstrTest, repeating_patterns) {
  EXPECT_EQ(solution.lengthOfLongestSubstring("abcabcbbabcabcbb"), 3);
  EXPECT_EQ(solution.lengthOfLongestSubstring("aaaaaaaa"), 1);
  EXPECT_EQ(solution.lengthOfLongestSubstring("abcdabcabcd"), 4);
}

File: JS/tests/longest_substr.test.js
import { test, expect } from 'vitest'
import { lengthOfLongestSubstring } from '../src/longest_substr.js'

test('should return 3 for the input "abcabcbb"', () => {
  const input = 'abcabcbb'
  const result = lengthOfLongestSubstring(input)
  expect(result).toBe(3)
})

test('should return 1 for the input "bbbbb"', () => {
  const input = 'bbbbb'
  const result = lengthOfLongestSubstring(input)
  expect(result).toBe(1)
})

test('should return 3 for the input "pwwkew"', () => {
  const input = 'pwwkew'
  const result = lengthOfLongestSubstring(input)
  expect(result).toBe(3)
})

test('should return 0 for an empty string', () => {
  const input = ''
  const result = lengthOfLongestSubstring(input)
  expect(result).toBe(0)
})

test('should return 1 for a single character string', () => {
  const input = 'a'
  const result = lengthOfLongestSubstring(input)
  expect(result).toBe(1)
})

test('should return 3 for the input "dvdf"', () => {
  const input = 'dvdf'
  const result = lengthOfLongestSubstring(input)
  expect(result).toBe(3)
})

test('should return 2 for the input "abba"', () => {
  const input = 'abba'
  const result = lengthOfLongestSubstring(input)
  expect(result).toBe(2)
})
